osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/PlayerGrid.cs

```diff
@@ -0,0 +1,167 @@
+                    Padding = new MarginPadding(player_spacing),
+                    Children = new Drawable[]
+                    {
+                        new Container
+                        {
+                            RelativeSizeAxes = Axes.Both,
+                            Child = facadeContainer = new FillFlowContainer<Facade>
+                            {
+                                RelativeSizeAxes = Axes.X,
+                                AutoSizeAxes = Axes.Y,
+                                Spacing = new Vector2(player_spacing),
+                            }
+                        },
+                        maximisedFacade = new Facade { RelativeSizeAxes = Axes.Both }
+                    }
+                },
+                cellContainer = new Container<Cell> { RelativeSizeAxes = Axes.Both }
+            };
+        }
+
+        /// <summary>
+        /// Adds a new cell with content to this grid.
+        /// </summary>
+        /// <param name="content">The content the cell should contain.</param>
+        /// <exception cref="InvalidOperationException">If more than 16 cells are added.</exception>
+        public void Add(Drawable content)
+        {
+            if (cellContainer.Count == 16)
+                throw new InvalidOperationException("Only 16 cells are supported.");
+
+            int index = cellContainer.Count;
+
+            var facade = new Facade();
+            facadeContainer.Add(facade);
+
+            var cell = new Cell(index, content) { ToggleMaximisationState = toggleMaximisationState };
+            cell.SetFacade(facade);
+
+            cellContainer.Add(cell);
+        }
+
+        /// <summary>
+        /// The content added to this grid.
+        /// </summary>
+        public IEnumerable<Drawable> Content => cellContainer.OrderBy(c => c.FacadeIndex).Select(c => c.Content);
+
+        // A depth value that gets decremented every time a new instance is maximised in order to reduce underlaps.
+        private float maximisedInstanceDepth;
+
+        private void toggleMaximisationState(Cell target)
+        {
+            // Iterate through all cells to ensure only one is maximised at any time.
+            foreach (var i in cellContainer.ToList())
+            {
+                if (i == target)
+                    i.IsMaximised = !i.IsMaximised;
+                else
+                    i.IsMaximised = false;
+
+                if (i.IsMaximised)
+                {
+                    // Transfer cell to the maximised facade.
+                    i.SetFacade(maximisedFacade);
+                    cellContainer.ChangeChildDepth(i, maximisedInstanceDepth -= 0.001f);
+                }
+                else
+                {
+                    // Transfer cell back to its original facade.
+                    i.SetFacade(facadeContainer[i.FacadeIndex]);
+                }
+            }
+        }
+
+        protected override void Update()
```

Comment:
Is there a reason for doing this in update rather than on adding/removing a new item?

Reply:
It could if it didn't use absolute sizing, sure.

Reply:
Ah, right, I see. Probably fine as-is, but was there a reason to do it this way? Complexity reduction?

Reply:
Mostly complexity reduction yep.